File: hackerrank/lambda-calculi-9/01_string_filtrator.clj
(defn reduce-string [bs [x & xs]]
  "leave only the first occurence for each character in input string"
  (cond (nil? x)         []
        (contains? bs x) (reduce-string bs xs)
        :else            (cons x
                               (reduce-string
                                (conj bs x) xs))))

(defn -main [] ((comp println #(apply str %) reduce-string) #{} (read-line)))

(-main)



;;;; considerations:
;;; check cond syntax
;;; check if there's an empty predicate
;;; check whether it's in or elem or whatever else
;;; look up destructuring
;;; Is -main necessary here?
;;; which sequence datatype is appropriate here? -> sets!


;;; paredit: how to change (f (g x) y) -> (f y (g x))? 
;;; emacs:   how to zap whitespace?

File: hackerrank/lambda-calculi-9/01_string_filtrator.hs
-- Enter your code here. Read input from STDIN. Print output to STDOUT
main = 
    getContents >>= putStrLn . (string_filtrator [])


string_filtrator _ [] = []
string_filtrator bs (x:xs) = if x `elem` bs then string_filtrator bs xs else x : string_filtrator (x:bs) xs
